pkg/adaptor/etcd/etcd.go

```diff
@@ -102,107 +93,81 @@ func (e *Etcd) SampleConfig() string {
+	if err != nil {
+		return err
 	}
-	return
+	for msg := range msgChan {
+		e.pipe.Send(msg)
+	}
+
+	log.With("path", e.path).Infoln("adaptor Start finished...")
+	return nil
 }
 
 // Listen starts the pipe's listener
 func (e *Etcd) Listen() (err error) {
+	log.With("path", e.path).Infoln("adaptor Listening...")
 	defer func() {
+		log.With("path", e.path).Infoln("adaptor Listen closing...")
 		e.pipe.Stop()
 	}()
 
-	return e.pipe.Listen(e.writeMessage, e.tableMatch)
+	return e.pipe.Listen(e.writeMessage, e.subKeyMatch)
 }
 
 // Stop the adaptor
 func (e *Etcd) Stop() error {
+	log.With("path", e.path).Infoln("adaptor Stopping...")
 	e.pipe.Stop()
-	e.stopped = true
+
+	close(e.doneChannel)
+	e.wg.Wait()
+
+	log.With("path", e.path).Infoln("adaptor Stopped")
 	return nil
 }
 
+func (e *Etcd) keyFilter(key string) bool {
+	return e.subKeyMatch.MatchString(key)
+}
+
 // writeMessage writes one message to the destination Postgres, or sends an error down the pipe
 // TODO this can be cleaned up.  I'm not sure whether this should pipe the error, or whether the
-//   caller should pipe the error
+// caller should pipe the error
 func (e *Etcd) writeMessage(msg message.Msg) (message.Msg, error) {
-	m, err := message.Exec(message.MustUseAdaptor("etcd").(etcd.Adaptor).UseClient(e.session), msg)
-	if err != nil {
-		e.pipe.Err <- adaptor.NewError(adaptor.ERROR, e.path, fmt.Sprintf("etcd error (%v)", err), msg.Data())
-	}
-
-	return m, err
-}
+	_, ns, _ := message.SplitNamespace(msg)
+	err := client.Write(e.client, e.writer, message.From(msg.OP(), fmt.Sprintf(".%s", ns), msg.Data()))
 
-// catdata pulls down the original tables
-func (e *Etcd) catData() error {
-	kc := client.NewKeysAPI(e.session)
-	resp, err := kc.Get(context.Background(), "/", &client.GetOptions{Recursive: true})
 	if err != nil {
-		return err
-	}
-	err = e.copyNode(resp.Node)
-	return err
-}
-
-func (e *Etcd) copyNode(n *client.Node) error {
-	_, err := e.writeMessage(message.MustUseAdaptor("etcd").From(ops.Insert, fmt.Sprintf("%s.%s", n.Key, n.String()), data.Data{n.Key: n.String()}))
-	for _, node := range n.Nodes {
-		err = e.copyNode(node)
-		if err != nil {
-			return err
-		}
+		e.pipe.Err <- adaptor.NewError(adaptor.ERROR, e.path, fmt.Sprintf("write message error (%s)", err), msg.Data)
 	}
-	return err
+	return msg, err
 }
 
-// tail the logical data
-func (e *Etcd) tailData() error {
-	kc := client.NewKeysAPI(e.session)
-	watcher := kc.Watcher("/", &client.WatcherOptions{Recursive: true})
-	for {
-		if e.stopped {
-			return nil
-		}
-		resp, err := watcher.Next(context.Background())
-		if err != nil {
-			return err
-		}
-		err = e.copyNode(resp.Node)
-		if err != nil {
-			return err
-		}
-	}
+// Config provides configuration options for a postgres adaptor
+// the notable difference between this and dbConfig is the presence of the Tail option
+type Config struct {
+	URI       string `json:"uri" doc:"the uri to connect to, in the form TODO"`
```

Comment:
Do these `doc` struct tags actually get used in the CLI? If so,  you can fix `TODO`